File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-3.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    goodNodes(root) {
        if (!root) return 0;

        let curr = root;
        let stack = [];
        let currentMax = root.val;
        let count = 0;

        while(curr || stack.length > 0){
            while (curr){
                console.log(`Node: ${curr.val}, Max: ${currentMax}`);
                
                if (curr.val >= currentMax) {
                    console.log("Increment");
                    count += 1
                }
                currentMax = Math.max(currentMax, curr.val);
                stack.push([curr, currentMax]);

                curr = curr.left;
            }
            [curr, currentMax] = stack.pop()
            curr = curr.right;
        }
        return count;
    }
}

File: Data Structures & Algorithms/binary-tree-diameter/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        def dfs(node):
            nonlocal res

            if not node:
                return 0

            right = dfs(node.right)
            left = dfs(node.left)
            res = max(res, right + left)

            return 1 + max(right, left)

        dfs(root)
        return res